### DISCUSSION TITLE
issue: Input field value is not re-rendered in controlled nested array when another array element is selected

### DISCUSSION THREAD

**Discussion by schiller-manuel** [react-hook-form/react-hook-form#11500] [Q&A]
### Version Number

7.50.1

### Codesandbox/Expo snack

https://codesandbox.io/p/sandbox/react-hook-form-conditional-forked-j3m6n9?file=%2Fsrc%2Findex.js

### Steps to reproduce

1. Go to [example on codesandbox](https://codesandbox.io/p/sandbox/react-hook-form-conditional-forked-j3m6n9?file=%2Fsrc%2Findex.js)
2. Switch from 0 to 1 in select
3. See that field values are not updated

see screen recording

https://github.com/react-hook-form/react-hook-form/assets/6340397/cb7714ce-10e8-4403-909d-a10750f761b5

### Expected behaviour

When the `selectedIndex` is changed from `0` to `1`, `name` changes from `test.0.values.${valueIndex}` to `test.1.values.${valueIndex}`. I would thus expect that the rendered input values change from `["aaa", "bbb", "ccc"]` to `["xxx", "yyy", "zzz"]`.

A workaround is to use the `name` as `key`, then the correct values are rendered:

```tsx
<li key={`test.${selectedIndex}.values.${valueIndex}`}>
```

https://github.com/react-hook-form/react-hook-form/assets/6340397/6f991d5c-fc85-4bca-a380-de30061459a1

### What browsers are you seeing the problem on?

Firefox, Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
could you update your CSB? it's missing.

**Comment by schiller-manuel** [comment]
sorry it was still in drafts, now it should be accessible

**Comment by bluebill1049** [comment]
the update would need to be a valid path for that registered controller input. `test.${selectedIndex}.values.${valueIndex}`, it doesn't work at your current implementation.

**Comment by schiller-manuel** [comment]
sorry I don't understand that. how would a valid path look like if mine is invalid?

**Comment by bluebill1049** [comment]
This is the input name which you have registered `test.${selectedIndex}.values.${valueIndex}`

and this is your setValue `onChange={(e) => setValue("selectedIndex", e.target.value)}`, setValue will need the specific controller name/path

**Comment by schiller-manuel** [comment]
`selectedIndex` is a field in the form state, next to `test`.

**Comment by bluebill1049** [comment]
> `selectedIndex` is a field in the form state, next to `test`.

yes, but it doesn't work like that. controller is not a state management input (for that entire object), it's just a input which store value for that particular path which you have giving by name/path.

you can't do this basically 


```tsx
name="test.test.test.test.test.test"
- setValue('test', { test: {test: {test: ''}}})
+ setValue('test.test.test.test.test.test', 'data')
```

**Reply by schiller-manuel** [reply]
I found the problem.

`useController` uses `useWatch` which does not react to `name` property change.
See:
- https://github.com/react-hook-form/react-hook-form/issues/9507 
- https://github.com/react-hook-form/react-hook-form/issues/9732
- https://github.com/react-hook-form/react-hook-form/issues/9727

I enhanced the example to also show the output from `useWatch` and how it does **not** get updated: [codesandbox](https://codesandbox.io/p/sandbox/react-hook-form-conditional-forked-cjjf4x?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsgwun5p00062e687fitrcfa%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsgwun5p00022e683dvc6b0f%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsgwun5p00032e681w7g44fy%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsgwun5p00052e68ktptuerh%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsgwun5p00022e683dvc6b0f%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsgwun5p00012e68ch9965hp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clsha5lrg00022e687p2c5a44%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A21%252C%2522startColumn%2522%253A5%252C%2522endLineNumber%2522%253A21%252C%2522endColumn%2522%253A5%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clsgwun5p00022e683dvc6b0f%2522%252C%2522activeTabId%2522%253A%2522clsgwun5p00012e68ch9965hp%2522%257D%252C%2522clsgwun5p00052e68ktptuerh%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsgwun5p00042e68z60hhts4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsgwun5p00052e68ktptuerh%2522%252C%2522activeTabId%2522%253A%2522clsgwun5p00042e68z60hhts4%2522%257D%252C%2522clsgwun5p00032e681w7g44fy%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clsgwun5p00032e681w7g44fy%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
